docs/contributor/commits_pull_requests_and_merging.md

```diff
@@ -0,0 +1,108 @@
+
+### Conventional Commits: Breaking Changes
+
+Annotating **BREAKING CHANGES** is extremely important to our release process and versioning. To mark a commit as breaking, we add the `!` character after the type, but before the colon. For example:
+
+```
+feat!: Update assignment semantics (#560)
+```
+
+### Conventional Commits: Scopes
+
+In the past, we weren't very disciplined about using scopes; however, they significantly improve the Changelog, so we want to use a scope whenever possible. If we are only changing one part of the project, we can use the name of the workspace, like `(stdlib)` or `(grainc)`. If a change touches multiple parts of the codebase, there might be a better scope, like `(refactor)`.
+
+```
+feat(stdlib): Add Array.reverse() function (#698)
+```
+
+```
+fix(grainc): Apply separator normalization before cmdliner util (#654)
+```
+
+## Pull Requests
+
+Pull requests should be focused on the specific change they are working towards. If additional work is required to finish the original pull request, that work should be submitted as a separate one.
+
+This strategy avoids scenarios where pull requests grow too large/out-of-scope and don't get proper reviews—we want to avoid "LGTM, I trust you" reviews.
+
+The easiest way to do this is to have multiple Conventional Commits while you work and then you can cherry-pick the smaller changes into separate branches for pull requesting.
+
+### Reviews
+
+As the Grain core team has grown, we've adjusted our code review process. Currently, we require **two** approvals from core team members for any non-breaking pull request and **three** approvals for pull requests containing breaking changes.
+
+### With Breaking Changes
+
+Sometimes, we don't merge pull requests with breaking changes immediately upon approval. Since a breaking change will cause Grain to bump to the next "minor" version, we might want to land some fixes in "patch" releases before we begin working on that next breaking version.
+
+## Merging
+
+Once approved by the required number of core team members, the submitter can merge their changes into the `main` branch. Sometimes, especially for external contributions, the final approver may merge the pull request instead of the submitter.
+
+We use "squash merging" to merge pull requests, this will cause all commits to be combined into one commit—another reason we want to keep pull requests small & focused.
+
+### Squash Merging
+
+When squash merging, we can keep intermediate Conventional Commits around by adding them to the body of the commit message; however, the GitHub UI adds a `*` character before each commit message and our releaser bot can't parse that.
+
+When squashing, you want to update both the title of the commit to be a good Conventional Commit and adjust the body to contain any other Conventional Commits that you want to keep (not prefixed with `*`). We also keep any "Co-Authored By:" lines at the bottom of the commit.
```

Comment:
I've started throwing out the `Co-authored-by`s introduced from PR suggestions, since once it's squashed that gives equal credit for the whole PR to the suggestor (versus a rebase where they'd only get credit for what they suggested). I don't wanna steal credit for a contributor's hard work! Any thoughts on that? Obviously not a huge deal but I thought I'd call it out.

Reply:
I think the only thing I'd want to make sure is to not lose a third contributor. How do you think I should note that?

Reply:
Ah, you mean not just doing it on autopilot? I think it's fine to say something about being mindful about it (and I'm pretty tired so I won't be offended if you need to completely reword/change it 😂):
```suggestion
When squashing, you want to update both the title of the commit to be a good Conventional Commit and adjust the body to contain any other Conventional Commits that you want to keep (not prefixed with `*`). We also keep any "Co-authored-by:" lines at the bottom of the commit if the change was done by multiple authors. If "Co-authored-by:" lines appear due to accepted PR suggestions, it's good to delete them so the author gets full credit for the change.

Our overall approach to squashing is to be mindful of the impact of the commit—it populates our changelog and thus it's important to properly convey what change has happened, but it's also a record that we and others will go back and look at in the future. Thus, it's important to attribute the change to its correct authors and provide useful information that future-us would want to have.
```